util/econftool.c

```diff
@@ -0,0 +1,557 @@
+  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+  SOFTWARE.
+*/
+
+#include <errno.h>
+#include <getopt.h>
+#include <stdio.h>
+#include <string.h>
+#include <unistd.h>
+#include <sys/wait.h>
+#include <limits.h>
+#include <ftw.h>
+
+#include "libeconf.h"
+
+static const char *utilname = "econftool";
+static econf_file *key_file = NULL;
+static bool non_interactive = false;
+char *suffix = NULL; /* the suffix of the filename e.g. .conf */
+char path[PATH_MAX]; /* the path of the config file */
+char filename[PATH_MAX]; /* the filename without the suffix */
+char filename_suffix[PATH_MAX]; /* the filename with the suffix */
+char path_filename[PATH_MAX]; /* the path concatenated with the filename and the suffix */
+char root_dir[PATH_MAX] = "/etc";
+char usr_root_dir[PATH_MAX] = "/usr/etc";
+char xdg_config_dir[PATH_MAX];
+
+/**
+ * @brief Shows the usage.
+ */
+
+static void usage(void)
+{
+    fprintf(stderr, "Usage: %s COMMANDS [OPTIONS] filename.conf\n\n", utilname);
+    fprintf(stderr, "COMMAND:\n");
+    fprintf(stderr, "show     reads all snippets for filename.conf and prints all groups,\n");
+    fprintf(stderr, "         keys and their values.\n");
+    fprintf(stderr, "edit     starts the editor $EDITOR (environment variable) where the\n");
+    fprintf(stderr, "         groups, keys and values can be modified and saved afterwards.\n");
+    fprintf(stderr, "  -f, --full:      copy the original configuration file to /etc instead of\n");
+    fprintf(stderr, "                   creating drop-in files.\n");
+    fprintf(stderr, "  -y, --yes:       Assumes yes for all prompts.\n");
+    fprintf(stderr, "revert   reverts all changes to the vendor versions. Basically deletes\n");
+    fprintf(stderr, "         the config file and snippet directory in /etc.\n");
+    fprintf(stderr, "  -y, --yes:       Assumes yes for all prompts and runs non-interactively.\n\n");
+}
+
+/**
+ * @brief Change root dir if enviroment variable "ECONFTOOL_ROOT" exists
+ *
+ * @param path The path to be changed
+ */
+static void change_root_dir(char path[PATH_MAX])
+{
+    if (getenv("ECONFTOOL_ROOT") != NULL) {
+        if (strlen(path) + strlen(getenv("ECONFTOOL_ROOT")) >= PATH_MAX) {
+            fprintf(stderr, "ECONFTOOL_ROOT + path is too long\n");
+            exit(EXIT_FAILURE);
+        }
+
+        char *tmp = strdup(path);
+
+        strcpy(path, getenv("ECONFTOOL_ROOT"));
+        strcat(path, tmp);
+
+        free(tmp);
+    }
+}
+
+/**
+ * @brief This function is used as a callback for nftw to delete
+ *        all files nftw finds.
+ */
+static int nftw_remove(const char *path, const struct stat *sb, int flag, struct FTW *ftwbuf)
+{
+    return remove(path);
+}
+
+/**
+ * @brief Generate a tmpfile using mkstemp() and write its name into tmp_name
+ */
+static int generate_tmp_file(char *tmp_name)
+{
+    int filedes;
+    char name_template[21] = "/tmp/econftool-XXXXXX";
```

Comment:
As you have a file name here anyway (`mkstemp` modifies the passed string), you could either just return that directly or alternatively use `tmpfile()` + `fileno` which produces an unnamed file.